lib/hsa/hsaRfiUpdateBadInputs.cpp

```diff
@@ -32,49 +29,77 @@ hsaRfiUpdateBadInputs::hsaRfiUpdateBadInputs(Config& config, const string& uniqu
     _network_buf_execute_id = 0;
     _network_buf_finalize_id = 0;
 
-    update_bad_inputs = false;
+    _in_buf = host_buffers.get_buffer("bad_inputs_buf");
+    register_consumer(_in_buf, unique_name.c_str());
+    _in_buf_precondition_id = 0;
+    first_pass = true;
+    num_bad_inputs = 0;
+
     frames_to_update = 0;
-    frames_to_update_finalize = 0;
+    frame_copy_active.insert(std::begin(frame_copy_active), device.get_gpu_buffer_depth(), false);
 
     // Alloc memory on GPU
     device.get_gpu_memory_array("input_mask", 0, input_mask_len);
-
-    kotekan::configUpdater::instance().subscribe(
-        config.get<std::string>(unique_name, "updatable_config/bad_inputs"),
-        std::bind(&hsaRfiUpdateBadInputs::update_bad_inputs_callback, this, std::placeholders::_1));
 }
 
 hsaRfiUpdateBadInputs::~hsaRfiUpdateBadInputs() {
     hsa_host_free(host_mask);
 }
 
+inline void hsaRfiUpdateBadInputs::copy_frame(int gpu_frame_id) {
+
+    frames_to_update = device.get_gpu_buffer_depth();
+    memcpy(host_mask, _in_buf->frames[_in_buf_precondition_id], input_mask_len);
+    num_bad_inputs = get_rfi_num_bad_inputs(_in_buf, _in_buf_precondition_id);
+    DEBUG("gpu_frame_id={:d} using _in_buf_precondition_id={:d}", gpu_frame_id,
+          _in_buf_precondition_id);
+    mark_frame_empty(_in_buf, unique_name.c_str(), _in_buf_precondition_id);
+    _in_buf_precondition_id = (_in_buf_precondition_id + 1) % _in_buf->num_frames;
+}
+
 int hsaRfiUpdateBadInputs::wait_on_precondition(int gpu_frame_id) {
     (void)gpu_frame_id;
 
-    uint8_t* frame =
-        wait_for_full_frame(_network_buf, unique_name.c_str(), _network_buf_precondition_id);
-    if (frame == nullptr)
-        return -1;
-
-    _network_buf_precondition_id = (_network_buf_precondition_id + 1) % _network_buf->num_frames;
+    // Check for bad input updates
+    std::lock_guard<std::mutex> lock(update_mutex);
+    if (first_pass) {
+        uint8_t* frame = wait_for_full_frame(_in_buf, unique_name.c_str(), _in_buf_precondition_id);
+        if (frame == NULL)
+            return -1;
+        first_pass = false;
+        copy_frame(gpu_frame_id);
+    } else {
+        // Check for new bad inputs only if all gpu frames have been updated (not currently updating
+        // frame)
+        if (frames_to_update == 0 && !frame_copy_active.at(gpu_frame_id)) {
```

Comment:
This actually should check all elements of `frame_copy_active()` are false.  Otherwise you could have a copy happening which hasn't yet finished, but `frames_to_update` would be zero, and you'd write into `host_mask` while it was being written.  See how I did this in #603